pkg/rpm/database.go

```diff
@@ -0,0 +1,285 @@
+package rpm
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/stackrox/rox/pkg/set"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/scanner/pkg/commonerr"
+	"github.com/stackrox/scanner/pkg/features"
+	"github.com/stackrox/scanner/pkg/tarutil"
+)
+
+const (
+	// This is the query format we're using to get data out of rpm.
+	queryFmt = `%{name}\n` +
+		`%{evr}\n` +
+		`%{ARCH}\n` +
+		`%{RPMTAG_MODULARITYLABEL}\n` +
+		`[%{FILENAMES}\n]` +
+		`.\n`
+
+	// Older versions of rpm do not have the `RPMTAG_MODULARITYLABEL` tag.
+	// Ignore it for testing.
+	queryFmtTest = `%{name}\n` +
+		`%{evr}\n` +
+		`%{ARCH}\n` +
+		`[%{FILENAMES}\n]` +
+		`.\n`
+
+	// databaseDir is the directory where the RPM database is expected to be in
+	// the container filesystem.
+	databaseDir = "var/lib/rpm"
+)
+
+var (
+	// databaseFiles is a set with all the supported RPM database files.
+	databaseFiles = set.NewStringSet(
+		// BerkleyDB (rpm < 4.16)
+		"Packages",
+	)
+)
+
+// rpmDatabase represents an RPM database in the filesystem.
+type rpmDatabase struct {
+	// The path to the rpm database, can be used as --dbpath <dbPath> in rpm commands.
+	dbPath string
+}
+
+// rpmDatabaseQuery is the RPM query iterator object.
+type rpmDatabaseQuery struct {
+	// RPM sub-process management and scanning.
+	rpmWait        func() error
+	rpmScanStopped bool
+	rpmScanner     *bufio.Scanner
+	rpmStderr      *bytes.Buffer
+	rpmStdout      io.ReadCloser
+	// Query state.
+	err         error
+	nextPackage rpmPackage
+}
+
+// rpmPackage represents an RPM package information.
+type rpmPackage struct {
+	Name      string
+	Version   string
+	Arch      string
+	Module    string
+	Filenames []string
+}
+
+// QueryOpts is a set of options for database queries.
+type QueryOpts struct {
```

Comment:
One, allowing callers to select the format gives them the power to break the parsing logic. I think the query format is an implementation detail. If we want to support customized query formats, we should create a better abstraction -- like formatting JSON in the rpm output, and unmarshal with optional fields, or something similar. This can be done in the future if needed, but not if we allow callers to pass the format string.

Two, using the struct allows callers to use the default value -- and pass down an empty struct `.QueryAll(QueryOpts{})` instead of using something like `.QueryAll("")`, which I think is easier to read and understand.  It also allows us to add options in the future without forcing callers to change their code -- which is aligned with the semantic of "optional parameters".